src/app/credExplorer/pagerankTable/aggregation.js

```diff
@@ -29,11 +33,131 @@ export type ConnectionAggregation = {|
+  const reduce = (totalScore, scoredConnection) =>
+    scoredConnection.connectionScore + totalScore;
+  return xs.reduce(reduce, 0);
+}
+
+export function aggregateByNodeType(
+  xs: $ReadOnlyArray<ScoredConnection>,
+  nodeTypes: $ReadOnlyArray<NodeType>
+): $ReadOnlyArray<NodeAggregation> {
+  const typeTrie = new NodeTrie();
+  for (const nodeType of nodeTypes) {
+    typeTrie.add(nodeType.prefix, nodeType);
+  }
+  const nodeTypeToConnections = new Map();
+  for (const x of xs) {
+    const types = typeTrie.get(x.source);
+    if (types.length === 0) {
+      throw new Error(
+        `No matching NodeType for ${NodeAddress.toString(x.source)}`
+      );
+    }
+    const type = types[types.length - 1];
+    if (!nodeTypeToConnections.has(type)) {
+      nodeTypeToConnections.set(type, []);
+    }
+    NullUtil.get(nodeTypeToConnections.get(type)).push(x);
+  }
+  const aggregations: NodeAggregation[] = [];
+  for (const [
+    nodeType: NodeType,
+    connections: ScoredConnection[],
+  ] of nodeTypeToConnections) {
+    const aggregation = {
+      nodeType,
+      connections: sortBy(connections, (x) => -x.connectionScore),
+      summary: {
+        size: connections.length,
+        score: sumScoredConnections(connections),
+      },
+    };
+    aggregations.push(aggregation);
+  }
+  return sortBy(aggregations, (x) => -x.summary.score);
+}
+
+type EdgeTypeToConnection = Map<EdgeType, ScoredConnection[]>;
 export function aggregateByConnectionType(
   xs: $ReadOnlyArray<ScoredConnection>,
   nodeTypes: $ReadOnlyArray<NodeType>,
   edgeTypes: $ReadOnlyArray<EdgeType>
 ): ConnectionAggregations {
-  const _unused_stuff = [xs, edgeTypes, nodeTypes];
-  throw new Error("Not yet implemented");
+  const typeTrie = new EdgeTrie();
+  for (const edgeType of edgeTypes) {
+    typeTrie.add(edgeType.prefix, edgeType);
+  }
+  const syntheticConnections = [];
+  const inEdgeTypeToConnections: EdgeTypeToConnection = new Map();
+  const outEdgeTypeToConnections: EdgeTypeToConnection = new Map();
+
+  for (const x of xs) {
+    let relevantMap: EdgeTypeToConnection;
+    switch (x.connection.adjacency.type) {
+      case "SYNTHETIC_LOOP":
+        syntheticConnections.push(x);
+        continue;
+      case "IN_EDGE":
+        relevantMap = inEdgeTypeToConnections;
+        break;
+      case "OUT_EDGE":
+        relevantMap = outEdgeTypeToConnections;
+        break;
+      default:
+        throw new Error((x.connection.adjacency.type: empty));
+    }
+    const edge = x.connection.adjacency.edge;
+    const types = typeTrie.get(edge.address);
+    if (types.length === 0) {
+      throw new Error(`No matching EdgeType for edge ${edgeToString(edge)}`);
+    }
+    const type = types[types.length - 1];
+    if (!relevantMap.has(type)) {
+      relevantMap.set(type, []);
+    }
+    NullUtil.get(relevantMap.get(type)).push(x);
+  }
+
+  const result = [];
+
+  if (syntheticConnections.length > 0) {
```

Comment:
This can all just be:

```js
  if (syntheticConnections.length > 0) {
    result.push(
      createAggregation({type: "SYNTHETIC_LOOP"}, syntheticConnections)
    );
  }
```

(in which case it should be below the fundef).

Reply:
Nice catch.